google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/migration/StagedArtifactNameConverterProvider.java

```diff
@@ -0,0 +1,124 @@
+import com.intellij.conversion.RunManagerSettings;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import org.jdom.Element;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A {@link ConverterProvider} for converting the staged artifact name in App Engine run
+ * configurations.
+ *
+ * <p>With the introduction of this field, existing run configurations may rely on the previous
+ * behavior for staging artifacts for custom runtime deployments. This behavior was to always rename
+ * the built artifact to "target.jar" or "target.war", depending on the type of artifact that was
+ * built.
+ *
+ * <p>This conversion flags all deployment configurations with a legacy bit that will be checked
+ * upon deployment. If set, the staged artifact name will be explicitly set to "target.jar" or
+ * "target.war" so the functionality is preserved.
+ */
+public final class StagedArtifactNameConverterProvider extends ConverterProvider {
+
+  public StagedArtifactNameConverterProvider() {
+    super("gcp-staged-artifact-name-converter");
+  }
+
+  @NotNull
+  @Override
+  public String getConversionDescription() {
+    return GctBundle.message("appengine.staged.artifact.name.converter.description");
+  }
+
+  @NotNull
+  @Override
+  public ProjectConverter createConverter(@NotNull ConversionContext context) {
+    return new ProjectConverter() {
+      @Override
+      public ConversionProcessor<RunManagerSettings> createRunConfigurationsConverter() {
+        return new StagedArtifactNameConversionProcessor(context);
+      }
+    };
+  }
+
+  @Override
+  public boolean canDetermineIfConversionAlreadyPerformedByProjectFiles() {
+    return false;
+  }
+
+  private static final class StagedArtifactNameConversionProcessor
+      extends ConversionProcessor<RunManagerSettings> {
+
+    static final Function<Element, Stream<Element>> TO_DEPLOYMENT_SETTINGS =
+        element -> {
+          Element deployment = element.getChild("deployment");
+          if (deployment != null) {
+            Element settings = deployment.getChild("settings");
+            if (settings != null) {
+              return Stream.of(settings);
+            }
+          }
+          return Stream.empty();
+        };
+
+    static final Predicate<Element> IS_GCP_APP_ENGINE_DEPLOY =
+        element -> "gcp-app-engine-deploy".equals(element.getAttributeValue("type"));
+
+    static final Predicate<Element> IS_LEGACY_BIT_NONEXISTENT =
+        element -> element.getAttribute(STAGED_ARTIFACT_NAME_LEGACY) == null;
+
+    final ConversionContext conversionContext;
+
+    StagedArtifactNameConversionProcessor(ConversionContext conversionContext) {
+      this.conversionContext = conversionContext;
+    }
+
+    @Override
+    public boolean isConversionNeeded(RunManagerSettings runManagerSettings) {
+      return runManagerSettings
+          .getRunConfigurations()
+          .stream()
+          .filter(IS_GCP_APP_ENGINE_DEPLOY)
```

Comment:
super nit: but some of this stream logic is duplicated with #process and perhaps could be extracted

Reply:
actually nm - I didn't notice the difference on IS_LEGACY_BIT_NONEXISTENT.